async-server.c

```diff
@@ -0,0 +1,727 @@
+{
+    struct srv_context_t *ctx = (struct srv_context_t *) context;
+
+    pthread_mutex_lock(&ctx->tasks_mutex);
+    ++ctx->tasks_running;
+    pthread_mutex_unlock(&ctx->tasks_mutex);
+
+    queue_push(&ctx->queue, task);
+    return (ctx->password[0] != 0);
+}
+
+static void *
+srv_server(void *arg)
+{
+    struct params_t *params = (struct params_t *) arg;
+    struct srv_context_t *context = (struct srv_context_t *) params->context;
+    int server_sfd = params->socket_fd;
+    int status;
+
+    while (true)
+    {
+        // Printing to stderr to be able to check output in tests
+        fprintf(stderr, "Waiting for new connections...\n");
+        int client_sfd = accept(server_sfd, NULL, NULL);
+        if (client_sfd == -1)
+            handle_error("accept");
+        fprintf(stderr, "Got new connection...\n");
+
+        int max_tasks;
+        status = recvall(client_sfd, &max_tasks, sizeof(max_tasks), 0);
+        if (status == -1)
+        {
+            close_client(client_sfd);
+            continue;
+        }
+        max_tasks *= 2;
+
+        pthread_mutex_lock(&context->set_mutex);
+        pthread_cleanup_push(
+            (void (*) (void*)) pthread_mutex_unlock,
+            &context->set_mutex
+        );
+
+        struct node_t *node = set_take_last(&context->set);
+        node->socket_fd = client_sfd;
+
+        node->handler_context = malloc(sizeof(*node->handler_context));
+        node->handler_context->max_tasks = max_tasks;
+        node->handler_context->current_tasks = 0;
+        node->handler_context->tasks =
+            malloc(max_tasks * sizeof(*node->handler_context->tasks));
+        id_set_init(&node->handler_context->id_set, max_tasks);
+        pthread_mutex_init(&node->handler_context->id_set_mutex, NULL);
+        node->handler_context->client_sfd = client_sfd;
+        node->handler_context->srv_context = context;
+
+        status = pthread_create(
+            &node->sender_id, NULL, task_sender, node->handler_context
+        );
+        if (status != 0)
+        {
+            perror("Couldn't start sender thread");
+            handler_context_destroy(node->handler_context);
+            free(node->handler_context);
+            close_client(client_sfd);
+        }
+
+        status = pthread_create(
+            &node->receiver_id, NULL, task_receiver, node->handler_context
+        );
+        if (status != 0)
+        {
+            perror("Couldn't start receiver thread");
+            handler_context_destroy(node->handler_context);
+            free(node->handler_context);
+            close_client(client_sfd);
+        }
+
+        pthread_cleanup_pop(!0);
+        pthread_mutex_unlock(&context->set_mutex);
+    }
+
+    return NULL;
+}
+
+bool
+run_async_server(struct task_t *task, struct config_t *config)
+{
+    struct srv_context_t context;
+    context.hash = config->hash;
+    context.tasks_running = 0;
+    pthread_mutex_init(&context.thread_started, NULL);
```

Comment:
It looks like this mutex is never used. It is required for synchronous thread creation and safe read of thread parameters.